File: api/app/store/conversation_evals.py
"""Persistence for conversation-eval Prompt versions, archives, and runs."""
import json

from app.store.observability import get_conn


def create_prompt_version(name: str, content: str) -> dict:
    with get_conn() as conn:
        cursor = conn.execute(
            "INSERT INTO conversation_eval_prompt_versions(name, content) VALUES (?, ?)",
            (name, content),
        )
        row = conn.execute(
            "SELECT * FROM conversation_eval_prompt_versions WHERE id = ?",
            (cursor.lastrowid,),
        ).fetchone()
    return dict(row)


def list_prompt_versions() -> list[dict]:
    with get_conn() as conn:
        rows = conn.execute(
            "SELECT * FROM conversation_eval_prompt_versions ORDER BY id DESC"
        ).fetchall()
    return [dict(row) for row in rows]


def get_prompt_version(version_id: int) -> dict | None:
    with get_conn() as conn:
        row = conn.execute(
            "SELECT * FROM conversation_eval_prompt_versions WHERE id = ?",
            (version_id,),
        ).fetchone()
    return dict(row) if row else None


_RECORD_FIELDS = (
    "source_user_turn", "source_assistant_turn", "stream", "source_created_at",
    "source_user_content", "baseline_output", "baseline_prompt_release_id",
    "baseline_prompt_release_version", "baseline_prompt_label",
    "baseline_system_prompt", "baseline_input_json", "prompt_kind",
    "prompt_release_id", "prompt_release_version", "prompt_version_id",
    "prompt_label", "system_prompt", "input_json", "runtime_snapshot_json",
)


def _record_row(row) -> dict:
    item = dict(row)
    for stored, exposed in (
        ("baseline_input_json", "baseline_input"),
        ("input_json", "input"),
        ("runtime_snapshot_json", "runtime_snapshot"),
    ):
        raw = item.pop(stored)
        item[exposed] = json.loads(raw) if raw else None
    return item


def create_record(record: dict) -> dict:
    with get_conn() as conn:
        cursor = conn.execute(
            "INSERT INTO conversation_eval_records("
            + ",".join(_RECORD_FIELDS) + ") VALUES ("
            + ",".join("?" for _ in _RECORD_FIELDS) + ")",
            tuple(record.get(field) for field in _RECORD_FIELDS),
        )
        row = conn.execute(
            "SELECT * FROM conversation_eval_records WHERE id = ?",
            (cursor.lastrowid,),
        ).fetchone()
    return _record_row(row)


def get_record(record_id: int) -> dict | None:
    with get_conn() as conn:
        row = conn.execute(
            "SELECT * FROM conversation_eval_records WHERE id = ?", (record_id,),
        ).fetchone()
    return _record_row(row) if row else None


def list_records(limit: int = 50, before: int | None = None) -> list[dict]:
    where = "WHERE r.id < ?" if before is not None else ""
    params = (before, limit) if before is not None else (limit,)
    with get_conn() as conn:
        rows = conn.execute(
            "SELECT r.id, r.source_user_turn, r.source_assistant_turn, r.stream, "
            "substr(r.source_user_content, 1, 500) AS source_user_content, "
            "substr(r.baseline_output, 1, 500) AS baseline_output, "
            "r.baseline_prompt_label, "
            "r.prompt_kind, r.prompt_release_id, r.prompt_release_version, "
            "r.prompt_version_id, r.prompt_label, r.created_at, "
            "(SELECT COUNT(*) FROM conversation_eval_runs run "
            " WHERE run.record_id = r.id) AS run_count, "
            "(SELECT status FROM conversation_eval_runs run "
            " WHERE run.record_id = r.id ORDER BY run.id DESC LIMIT 1) AS latest_status, "
            "(SELECT substr(output, 1, 500) FROM conversation_eval_runs run "
            " WHERE run.record_id = r.id ORDER BY run.id DESC LIMIT 1) AS latest_output "
            "FROM conversation_eval_records r " + where + " ORDER BY r.id DESC LIMIT ?",
            params,
        ).fetchall()
    return [dict(row) for row in rows]


_RUN_FIELDS = (
    "record_id", "source_user_turn", "source_assistant_turn", "stream",
    "source_user_content", "original_content", "prompt_kind",
    "prompt_release_id", "prompt_release_version", "prompt_version_id",
    "prompt_label", "system_prompt", "input_json", "model",
)


def create_run(record: dict) -> int:
    with get_conn() as conn:
        cursor = conn.execute(
            "INSERT INTO conversation_eval_runs(" + ",".join(_RUN_FIELDS) + ") "
            "VALUES (" + ",".join("?" for _ in _RUN_FIELDS) + ")",
            tuple(record.get(field) for field in _RUN_FIELDS),
        )
        return cursor.lastrowid


def finish_run(
    run_id: int, *, output: str, reasoning: str | None,
    tool_calls: list[dict] | None, prompt_tokens: int | None,
    completion_tokens: int | None, duration_ms: int | None,
) -> None:
    with get_conn() as conn:
        conn.execute(
            "UPDATE conversation_eval_runs SET status = 'done', output = ?, "
            "reasoning = ?, tool_calls_json = ?, prompt_tokens = ?, "
            "completion_tokens = ?, duration_ms = ?, error = NULL, "
            "finished_at = datetime('now') WHERE id = ?",
            (
                output, reasoning,
                json.dumps(tool_calls, ensure_ascii=False) if tool_calls else None,
                prompt_tokens, completion_tokens, duration_ms, run_id,
            ),
        )


def fail_run(run_id: int, error: str) -> None:
    with get_conn() as conn:
        conn.execute(
            "UPDATE conversation_eval_runs SET status = 'error', error = ?, "
            "finished_at = datetime('now') WHERE id = ?",
            (error, run_id),
        )


def _run_row(row) -> dict:
    item = dict(row)
    raw_tools = item.pop("tool_calls_json")
    item["tool_calls"] = json.loads(raw_tools) if raw_tools else None
    # The record owns the immutable Prompt/input snapshot; result cards stay small.
    for field in (
        "system_prompt", "input_json", "source_user_content",
        "original_content", "reasoning",
    ):
        item.pop(field, None)
    return item


def get_run(run_id: int) -> dict | None:
    with get_conn() as conn:
        row = conn.execute(
            "SELECT * FROM conversation_eval_runs WHERE id = ?", (run_id,),
        ).fetchone()
    return _run_row(row) if row else None


def runs_for_record(record_id: int) -> list[dict]:
    with get_conn() as conn:
        rows = conn.execute(
            "SELECT * FROM conversation_eval_runs WHERE record_id = ? ORDER BY id DESC",
            (record_id,),
        ).fetchall()
    return [_run_row(row) for row in rows]

File: web/src/api/conversationEvals.ts
import { splitFrames } from "./sse";

export type ConversationEvalRound = {
  user_turn: number;
  assistant_turn: number;
  stream: string;
  user_content: string;
  original_content: string;
  created_at: string;
  trace_id: number | null;
  prompt_release_id: number | null;
  prompt_release_version: number | null;
  model: string | null;
  replayable: boolean;
};

export type ConversationPromptRelease = {
  id: number;
  version: number;
  note: string;
  published_at: string;
  is_active: boolean;
};

export type ConversationPromptVersion = {
  id: number;
  name: string;
  content: string;
  created_at: string;
};

export type ConversationModelCandidate = {
  id: string;
  name: string;
  model: string;
  configured: boolean;
};

export type ConversationEvalRun = {
  id: number;
  record_id: number | null;
  source_user_turn: number;
  source_assistant_turn: number;
  stream: string;
  prompt_kind: "original" | "release" | "custom";
  prompt_release_id: number | null;
  prompt_release_version: number | null;
  prompt_version_id: number | null;
  prompt_label: string;
  model: string | null;
  status: "running" | "done" | "error";
  output: string | null;
  error: string | null;
  prompt_tokens: number | null;
  completion_tokens: number | null;
  duration_ms: number | null;
  created_at: string;
  finished_at: string | null;
};

export type ConversationEvalWorkspace = {
  model_candidates: ConversationModelCandidate[];
  default_model_candidate: string;
  rounds: ConversationEvalRound[];
  releases: ConversationPromptRelease[];
  prompt_versions: ConversationPromptVersion[];
  has_more: boolean;
};

export type ConversationEvalRoundDetail = {
  round: ConversationEvalRound;
  original_system_prompt: string | null;
};

export type ConversationPromptMessage = {
  role: string;
  content: string;
  name?: string;
};

export type ConversationEvalRecordSummary = {
  id: number;
  source_user_turn: number;
  source_assistant_turn: number;
  stream: string;
  source_user_content: string;
  baseline_output: string;
  baseline_prompt_label: string;
  prompt_kind: "original" | "release" | "custom";
  prompt_release_id: number | null;
  prompt_release_version: number | null;
  prompt_version_id: number | null;
  prompt_label: string;
  run_count: number;
  latest_status: ConversationEvalRun["status"] | null;
  latest_output: string | null;
  created_at: string;
};

export type ConversationEvalRecordDetail = ConversationEvalRecordSummary & {
  source_created_at: string | null;
  baseline_prompt_release_id: number | null;
  baseline_prompt_release_version: number | null;
  baseline_system_prompt: string | null;
  baseline_input: ConversationPromptMessage[] | null;
  system_prompt: string;
  input: ConversationPromptMessage[];
  runs: ConversationEvalRun[];
};

export type ConversationEvalRecordPage = {
  records: ConversationEvalRecordSummary[];
  has_more: boolean;
};

export type ConversationEvalRunRequest = {
  assistant_turn: number;
  prompt_kind: "original" | "release" | "custom";
  prompt_release_id?: number;
  prompt_version_id?: number;
  model_candidate?: string;
};

type ConversationEvalRecordRunRequest = {
  model_candidate: string;
};

async function responseError(response: Response, action: string): Promise<Error> {
  try {
    const body = await response.json() as { detail?: unknown };
    if (typeof body.detail === "string" && body.detail) return new Error(body.detail);
  } catch {
    // Status fallback below handles empty and non-JSON responses.
  }
  return new Error(`${action} failed: ${response.status}`);
}

export async function getConversationEvalWorkspace(
  opts: { limit?: number; before?: number } = {},
): Promise<ConversationEvalWorkspace> {
  const query = new URLSearchParams({ limit: String(opts.limit ?? 50) });
  if (opts.before !== undefined) query.set("before", String(opts.before));
  const response = await fetch(`/inspect/conversation-evals?${query}`, {
    cache: "no-store",
  });
  if (!response.ok) throw await responseError(response, "load conversation evals");
  return response.json();
}

export async function getConversationEvalRound(
  assistantTurn: number,
): Promise<ConversationEvalRoundDetail> {
  const response = await fetch(
    `/inspect/conversation-evals/rounds/${assistantTurn}`,
    { cache: "no-store" },
  );
  if (!response.ok) throw await responseError(response, "load conversation round");
  return response.json();
}

export async function getConversationEvalRecords(
  opts: { limit?: number; before?: number } = {},
): Promise<ConversationEvalRecordPage> {
  const query = new URLSearchParams({ limit: String(opts.limit ?? 50) });
  if (opts.before !== undefined) query.set("before", String(opts.before));
  const response = await fetch(`/inspect/conversation-evals/records?${query}`, {
    cache: "no-store",
  });
  if (!response.ok) throw await responseError(response, "load evaluation archives");
  return response.json();
}

export async function getConversationEvalRecord(
  recordId: number,
): Promise<ConversationEvalRecordDetail> {
  const response = await fetch(
    `/inspect/conversation-evals/records/${recordId}`,
    { cache: "no-store" },
  );
  if (!response.ok) throw await responseError(response, "load evaluation archive");
  return response.json();
}

export async function createConversationEvalRecord(
  request: ConversationEvalRunRequest,
): Promise<ConversationEvalRecordDetail> {
  const response = await fetch("/inspect/conversation-evals/records", {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(request),
  });
  if (!response.ok) throw await responseError(response, "create evaluation archive");
  return response.json();
}

export async function createConversationPromptVersion(
  name: string,
  content: string,
): Promise<ConversationPromptVersion> {
  const response = await fetch("/inspect/conversation-evals/prompt-versions", {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ name, content }),
  });
  if (!response.ok) throw await responseError(response, "save Prompt version");
  return response.json();
}

type RunHandlers = {
  onRun?: (run: ConversationEvalRun) => void;
  onDelta?: (text: string) => void;
  onActivity?: (activity: Record<string, unknown>) => void;
  onDone?: (run: ConversationEvalRun) => void;
};

function parseFrame(frame: string): {
  kind: "run" | "delta" | "activity" | "done" | "end";
  data?: unknown;
} | null {
  const line = frame.split("\n").find((part) => part.startsWith("data:"));
  if (!line) return null;
  const payload = line.slice(5).trim();
  if (payload === "[DONE]") return { kind: "end" };
  try {
    const value = JSON.parse(payload) as Record<string, unknown>;
    if (value.run) return { kind: "run", data: value.run };
    if (value.delta) return { kind: "delta", data: value.delta };
    if (value.activity) return { kind: "activity", data: value.activity };
    if (value.done) return { kind: "done", data: value.done };
  } catch {
    return null;
  }
  return null;
}

async function streamConversationEvalAt(
  url: string,
  request: ConversationEvalRunRequest | ConversationEvalRecordRunRequest | null,
  handlers: RunHandlers,
  opts: { idleTimeoutMs?: number } = {},
): Promise<void> {
  const controller = new AbortController();
  const response = await fetch(url, {
    method: "POST",
    ...(request ? {
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(request),
    } : {}),
    signal: controller.signal,
  });
  if (!response.ok || !response.body) {
    throw await responseError(response, "run conversation replay");
  }

  const timeout = opts.idleTimeoutMs ?? 120_000;
  const reader = response.body.getReader();
  const decoder = new TextDecoder();
  let buffer = "";
  try {
    for (;;) {
      let timer: ReturnType<typeof setTimeout> | undefined;
      const read = reader.read();
      const deadline = new Promise<never>((_, reject) => {
        timer = setTimeout(() => {
          controller.abort();
          reject(new Error(`conversation replay timed out after ${timeout}ms`));
        }, timeout);
      });
      const { value, done } = await Promise.race([read, deadline]).finally(
        () => clearTimeout(timer),
      );
      if (done) break;
      buffer += decoder.decode(value, { stream: true });
      const split = splitFrames(buffer);
      buffer = split.rest;
      for (const raw of split.frames) {
        const frame = parseFrame(raw);
        if (!frame) continue;
        if (frame.kind === "end") return;
        if (frame.kind === "run") {
          handlers.onRun?.(frame.data as ConversationEvalRun);
        } else if (frame.kind === "delta") {
          const data = frame.data as { text?: unknown };
          if (typeof data.text === "string") handlers.onDelta?.(data.text);
        } else if (frame.kind === "activity") {
          handlers.onActivity?.(frame.data as Record<string, unknown>);
        } else if (frame.kind === "done") {
          handlers.onDone?.(frame.data as ConversationEvalRun);
        }
      }
    }
  } finally {
    controller.abort();
  }
}

export async function streamConversationEvalRun(
  request: ConversationEvalRunRequest,
  handlers: RunHandlers,
  opts: { idleTimeoutMs?: number } = {},
): Promise<void> {
  return streamConversationEvalAt(
    "/inspect/conversation-evals/run", request, handlers, opts,
  );
}

export async function streamConversationEvalRecordRun(
  recordId: number,
  modelCandidate: string,
  handlers: RunHandlers,
  opts: { idleTimeoutMs?: number } = {},
): Promise<void> {
  return streamConversationEvalAt(
    `/inspect/conversation-evals/records/${recordId}/runs`,
    { model_candidate: modelCandidate },
    handlers,
    opts,
  );
}
